#' Extract mutational signatures from trinucletide context.
#'
#' @description  Extracts adjacent bases to mutation site and classifies them into 96 substitution classes based on trinucleotide context followed by NMF decomposition to extract \code{n} signatures.
#'
#' @details This function loads reference genome into memeory. Typical human geneome occupies a peak memory of ~3 gb while extracting bases. NMF decomposition uses \code{nmf} function from NMF package.
#' Extracted signatures are compared against 21 experimentally validated signatures by calculating correlation coefficient. See http://www.nature.com/nature/journal/v500/n7463/fig_tab/nature12477_F2.html for details.
#'
#' @param maf an MAF object generated by \code{read.maf}
#' @param ref_genome faidx indexed refrence fasta file.
#' @param prefix Prefix to add or remove from contig names in MAF file.
#' @param add If prefix is used, default is to add prefix to contig names in MAF file. If false prefix will be removed from contig names.
#' @param ignoreChr Chromsomes to remove from analysis. e.g. chrM
#' @param n decompose matrix into n signatures. Default 3 signatures.
#' @return returns a list with signatures and 96 substitution matrix.
#' @export


ExtractSignatures = function(maf, ref_genome, prefix = NULL, add = TRUE, ignoreChr = NULL, n = 3){

  suppressPackageStartupMessages(require('VariantAnnotation', quietly = T))
  suppressPackageStartupMessages(require('Biostrings', quietly = T))

  maf = maf$data

  #one bp up and down.
  up = down = 1

  #reate a reference to an indexed fasta file.
  ref = FaFile(file = ref_genome)

  if(!is.null(prefix)){
    if(add){
      maf$Chromosome = paste(prefix,maf$Chromosome, sep = '')
    }else{
      maf$Chromosome = gsub(pattern = prefix, replacement = '', x = maf$Chromosome, fixed = T)
    }
  }

  #Remove unwanted contigs
  if(!is.null(ignoreChr)){
    maf = maf[!maf$Chromosome %in% ignoreChr]
  }

  #Some TCGA studies have Start_Position set to as 'position'. Change if so.
  if(length(grep(pattern = 'Start_position', x = colnames(maf))) > 0){
    colnames(maf)[which(colnames(maf) == 'Start_position')] = 'Start_Position'
  }

  if(length(grep(pattern = 'End_position', x = colnames(maf))) > 0){
    colnames(maf)[which(colnames(maf) == 'Start_position')] = 'End_Position'
  }

  #seperate snps and indels
  maf.snp = maf[Variant_Type == 'SNP']
  #maf.rest = maf[!maf$Variant_Type %in% 'SNP']

  #get unique Chromosome names from maf
  chrs = unique(maf.snp$Chromosome)

  #read fasta
  message('reading fasta..')
  ref = getSeq(x = ref)

  #extract contigs from reference fasta
  seq.lvl = seqlevels(ref)

  chrs.missing = chrs[!chrs %in% seq.lvl]

  #validation
  if(length(chrs.missing) > 0){
    message("contigs in fasta file:")
    print(seq.lvl)
    message("contigs in maf:")
    print(chrs)
    message('missing reference contigs from fasta file.')
    print(chrs.missing)
    stop("Contig names in maf must match to contig names in reference fasta. Use prefix to add or remove prefixes from maf if its necessary. Use ignoreChr to ignore missing contigs.")
  }


  extract.tbl = data.table(Chromosome = maf.snp$Chromosome, Start = maf.snp$Start_Position-1, End = maf.snp$Start_Position+1,
                          Reference_Allele = maf.snp$Reference_Allele, Tumor_Seq_Allele2 = maf.snp$Tumor_Seq_Allele2, Tumor_Sample_Barcode = maf.snp$Tumor_Sample_Barcode)

  message('Extracting trinucleotide context..')
  ss = subseq(x = ref[extract.tbl[,Chromosome]], start = extract.tbl[,Start] , end = extract.tbl[,End])
  extract.tbl[,trinucleotide:= as.character(ss)]
  extract.tbl[,Substitution:=paste(extract.tbl$Reference_Allele, extract.tbl$Tumor_Seq_Allele2, sep='>')]

  suppressWarnings( extract.tbl[,SubstitutionType:= as.character(factor(x = extract.tbl$Substitution, levels = c('A>G', 'T>C', 'C>T', 'G>A', 'A>T', 'T>A', 'A>C', 'T>G', 'C>A', 'G>T', 'C>G', 'G>C'),
         labels = c('T>C', 'T>C', 'C>T', 'C>T', 'T>A', 'T>A', 'T>G', 'T>G', 'C>A', 'C>A', 'C>G', 'C>G')))])

  type = paste(substr(x = as.character(extract.tbl$trinucleotide), 1, 1),'[',extract.tbl$SubstitutionType, ']', substr(as.character(extract.tbl$trinucleotide), 3, 3), sep='')
  extract.tbl[,SomaticMutationType:= type]

  extract.tbl.summary = extract.tbl[,.N , by = list(Tumor_Sample_Barcode, SomaticMutationType)]

  colOrder = c("A[C>A]A", "A[C>A]C", "A[C>A]G", "A[C>A]T", "C[C>A]A", "C[C>A]C",
    "C[C>A]G", "C[C>A]T", "G[C>A]A", "G[C>A]C", "G[C>A]G", "G[C>A]T",
    "T[C>A]A", "T[C>A]C", "T[C>A]G", "T[C>A]T", "A[C>G]A", "A[C>G]C",
    "A[C>G]G", "A[C>G]T", "C[C>G]A", "C[C>G]C", "C[C>G]G", "C[C>G]T",
    "G[C>G]A", "G[C>G]C", "G[C>G]G", "G[C>G]T", "T[C>G]A", "T[C>G]C",
    "T[C>G]G", "T[C>G]T", "A[C>T]A", "A[C>T]C", "A[C>T]G", "A[C>T]T",
    "C[C>T]A", "C[C>T]C", "C[C>T]G", "C[C>T]T", "G[C>T]A", "G[C>T]C",
    "G[C>T]G", "G[C>T]T", "T[C>T]A", "T[C>T]C", "T[C>T]G", "T[C>T]T",
    "A[T>A]A", "A[T>A]C", "A[T>A]G", "A[T>A]T", "C[T>A]A", "C[T>A]C",
    "C[T>A]G", "C[T>A]T", "G[T>A]A", "G[T>A]C", "G[T>A]G", "G[T>A]T",
    "T[T>A]A", "T[T>A]C", "T[T>A]G", "T[T>A]T", "A[T>C]A", "A[T>C]C",
    "A[T>C]G", "A[T>C]T", "C[T>C]A", "C[T>C]C", "C[T>C]G", "C[T>C]T",
    "G[T>C]A", "G[T>C]C", "G[T>C]G", "G[T>C]T", "T[T>C]A", "T[T>C]C",
    "T[T>C]G", "T[T>C]T", "A[T>G]A", "A[T>G]C", "A[T>G]G", "A[T>G]T",
    "C[T>G]A", "C[T>G]C", "C[T>G]G", "C[T>G]T", "G[T>G]A", "G[T>G]C",
    "G[T>G]G", "G[T>G]T", "T[T>G]A", "T[T>G]C", "T[T>G]G", "T[T>G]T"
    )

  #colOrderClasses = rep(c('C>A', 'C>G', 'C>T', 'T>A', 'T>C', 'T>G'), each = 16)

  conv.mat = as.data.frame(cast(extract.tbl.summary, formula = Tumor_Sample_Barcode~SomaticMutationType, fill = 0, value = 'N'))
  #head(conv.mat)
  rownames(conv.mat) = conv.mat[,1]
  #head(conv.mat)
  conv.mat = conv.mat[,-1]

  #conv.mat = t(t(conv.mat)[colOrder,])
  #Check for missing somatic mutation types (this is possible for cancer types with low mutation rate or for a cohort with lesser samples)
  colOrder.missing = colOrder[!colOrder %in% colnames(conv.mat)]
  #If any missing add them with zero counts
  if(length(colOrder.missing) > 0){
    zeroMat = as.data.frame(matrix(data = 0, nrow = nrow(conv.mat), ncol = length(colOrder.missing)))
    colnames(zeroMat) = colOrder.missing
    conv.mat = cbind(conv.mat, zeroMat)
  }

  conv.mat = as.matrix(conv.mat[,match(colOrder, colnames(conv.mat))]) #organize columns according to colOrder

  #Set NAs to zeros if any
  conv.mat[is.na(conv.mat)] = 0

  #Notes:
  #Available methods for nmf decompositions are 'brunet', 'lee', 'ls-nmf', 'nsNMF', 'offset'.
  #But based 21 breast cancer signatures data, defualt brunet seems to be working close to the results.
  #Sticking with default for now.

  conv.mat.nmf = nmf(x = t(conv.mat), rank = n)

  conv.mat.nmf.signatures = basis(conv.mat.nmf)
  conv.mat.nmf.signatures = apply(conv.mat.nmf.signatures, 2, function(x) x/sum(x)) #Scale the signatures

  colnames(conv.mat.nmf.signatures) = paste('Signature', 1:ncol(conv.mat.nmf.signatures),sep='_')
  #conv.mat.nmf.signatures.melted = melt(conv.mat.nmf.signatures)
  #levels(conv.mat.nmf.signatures.melted$X1) = colOrder

  sigs = fread(input = system.file('extdata', 'signatures.txt', package = 'maftools'), stringsAsFactors = F, data.table = F)
  colnames(sigs) = gsub(pattern = ' ', replacement = '_', x = colnames(sigs))
  rownames(sigs) = sigs$Somatic_Mutation_Type
  sigs = sigs[,-c(1:3)]
  sigs = sigs[,1:22] #use only first 21 validated sigantures
  sigs = sigs[rownames(conv.mat.nmf.signatures),]

  ###Plot
  plotData = as.data.frame(t(conv.mat.nmf.signatures))
  nsigs = nrow(plotData)
  par(mfrow = c(nsigs,1),oma = c(5,4,0,0) + 0.1,mar = c(0,0,1,1) + 0.1)
  color = rep(c("blue","black","red","gray","green","maroon"),each=16)

  for(i in 1:nsigs){

    barplot2(as.matrix(plotData[i,]),col=color,beside = T,axisnames=F,ylim=c(-0.1,0.3),cex.axis = 0.6,axes=F)
    axis(side = 2,at = c(0,0.05,0.1,0.15,0.2,0.25),labels = c(0,0.05,0.1,0.15,0.2,0.25),pos = c(0,0.05,0.1,0.15,0.2,0.25),cex.axis=0.8)
    abline(h = c(0.05,0.1,0.15,0.2,0.25),lty=2,lwd=0.3)
    text(labels = c("C>A","C>G","C>T","T>A","T>C","T>G"),y = rep(-0.05,6),x = c(20,50,80,110,140,170))

  }

  message('Comparing against experimentally validated 21 signatures.. (See Alexandrov et.al Nature 2013 for details.)')
  corMat = c()
  for(i in 1:ncol(conv.mat.nmf.signatures)){
    sig = conv.mat.nmf.signatures[,i]
    corMat = rbind(corMat, apply(sigs, 2, function(x) cor.test(x, sig)$estimate[[1]])) #Calulate correlation coeff.
  }
  rownames(corMat) = colnames(conv.mat.nmf.signatures)

  for(i in 1:nrow(corMat)){
    message(rownames(corMat)[i], ' most similar to ',names(which(corMat[i,] == max(corMat[i,]))), '. Correlation coeff: ', max(corMat[i,]), sep=' ')
  }

  return(list(signatures = conv.mat.nmf.signatures, matrix = conv.mat, corTable = corMat))
}
