#' Extract bases flanking the mutation.
#'
#' @description  Extract adjacent bases to mutation site. Defaults to one base immediate to 5' and 3' of Start_position.
#' Make sure chromosome names in 'Chromosome' field of maf matches to chromosome names in reference fasta file. This is useful for analyzing mutational signatures.
#'
#' @details This function loads reference genome into memeory. Typical human geneome occupies a peak memory of ~3 gb while extracting bases.
#'
#' @param maf output generated by \code{read.maf}
#' @param ref_genome faidx indexed refrence fasta file.
#' @param up how many bases upstream of Start_position. Defaults to 1
#' @param down how many bases downstream of Start_position. Defaults to 1
#' @return adds an extra column 'bases' to data slot of \code{maf}
#' @export


addBases = function(maf, ref_genome, up = 1, down = 1){

  require('VariantAnnotation', quietly = T)
  require('Biostrings', quietly = T)

  maf = maf$data

  #reate a reference to an indexed fasta file.
  ref = FaFile(file = ref_genome)

  #seperate snps and indels
  maf.snp = maf[maf$Variant_Type %in% 'SNP']
  maf.rest = maf[!maf$Variant_Type %in% 'SNP']

  #get unique Chromosome names from maf
  chrs = unique(maf.snp$Chromosome)

  #read fasta
  message('reading fasta..')
  ref = getSeq(x = ref)

  #extract contigs from reference fasta
  seq.lvl = seqlevels(ref)
  chrs.missing = chrs[!chrs %in% seq.lvl]

  #validation
  if(length(chrs.missing) > 0){
    message("contigs in fasta file:")
    print(seq.lvl)
    message("contigs in maf:")
    print(chrs)
    message('missing reference contigs from fasta file.')
    print(chrs.missing)
    stop("Contig names in maf must match to contig names in reference fasta.")
  }

  #empty vector to hold bases
  bases = c()

  message(paste('extracting',up, 'bases from upstream and', down, 'bases from downstream..'))
  #text progress bar
  pb = txtProgressBar(min = 0, max = nrow(maf.snp), style = 3)

  for(i in 1:nrow(maf.snp)){
    setTxtProgressBar(pb, i)
    locus = maf.snp[i,Start_Position] #locus
    chrom = maf.snp[i,Chromosome] #chromsome
    bases = c(bases, toString(subseq(x = ref[chrom], start = locus - up, end = locus + down)[[1]])) #extract bases from reference and add it to 'bases' holder
  }

  #add column bases to maf
  maf.snp[,bases := bases]
  maf.rest[,bases := NA]
  maf.tot = rbind(maf.snp, maf.rest)
  maf$data = maf.tot #replace maf slot with updated maf
  return(maf)
}
