#' add read counts to MAF file from bam files.
#' @description  add read counts to MAF file from bam files.
#'
#' @details This is a wrapper script which runs bam-readcount programme via R \code{\link[base]{system()}} command to get readcounts.
#' Assumes bam-readcount is installed and under path.
#' Make sure chromosome names in 'Chromosome' field of maf matches to chromosome names in reference fasta file.
#' Currently only supports for SNPs. Indels will be populated with NA.
#'
#' @param maf an MAF object generated by \code{\link{read.maf}}
#' @param bam_df two column dataframe. First column must contain Tumor_Sample_Barcode and second column with path to corresponding bam file.
#' @param ref_genome path to reference genome. required.
#' @param skipSamples Tumor_Sample_Barcodes to skip from bam_df.
#' @param samples add read counts to only these samples.
#' @param prefix Prefix to add or remove from contig names in MAF file.
#' @param add If prefix is used, default is to add prefix to contig names in MAF file. If false prefix will be removed from contig names.
#' @return maf file with "t_depth", "t_ref_count", "t_alt_count" and "t_vaf" columns added.
#' @source bam-readcount: https://github.com/genome/bam-readcount
#' @export

addReadCounts = function(maf, bam_df, BaseQuality = 10, MapQuality = 10, ref_genome = NULL, skipSamples = NULL, samples = NULL, prefix = NULL, add = TRUE){

  #check if bam-readcount is installed and under path
  if(Sys.which('bam-readcount') == ''){
    stop('bam-readcount not found. Make sure its installed and under path.\n source: https://github.com/genome/bam-readcount')
  }

  if(is.null(ref_genome)){
    message("ref_genome required !")
    stop(call. = T)
  }

  maf.dat = maf@data
  
  #Some TCGA studies have Start_Position set to as 'position'. Change if so.
  if(length(grep(pattern = 'Start_position', x = colnames(maf.dat))) > 0){
    colnames(maf.dat)[which(colnames(maf.dat) == 'Start_position')] = 'Start_Position'
  }
  
  if(length(grep(pattern = 'End_position', x = colnames(maf.dat))) > 0){
    colnames(maf.dat)[which(colnames(maf.dat) == 'End_position')] = 'End_Position'
  }
  
  if(!is.null(prefix)){
    if(add){
      maf.dat$Chromosome = paste(prefix,maf.dat$Chromosome, sep = '')
    }else{
      maf.dat$Chromosome = gsub(pattern = prefix, replacement = '', x = maf.dat$Chromosome, fixed = T)
    }
  }

  if(!is.null(skipSamples)){
    maf.dat = maf.dat[!maf.dat$Tumor_Sample_Barcode %in% skipSamples,]
  }

  if(!is.null(samples)){
    maf.dat = maf.dat[maf.dat$Tumor_Sample_Barcode %in% samples,]
  }

  tsbs = levels(factor(as.character(maf.dat$Tumor_Sample_Barcode)))
  #maf.split = split(maf, as.factor(as.character(maf$Tumor_Sample_Barcode)))
  maf.df = c()

  for(i in 1:length(tsbs)){

    tsb = tsbs[i]

    maf = data.frame(maf.dat[Tumor_Sample_Barcode == tsbs[i]])
    maf = maf[with(maf, order(Chromosome, Start_Position)),]
    maf$id = paste(as.character(maf$Chromosome), as.character(maf$Start_Position), sep = ':')
    maf.dups = maf[duplicated(maf$id),]

    if(nrow(maf.dups) > 0){
      message('Duplicated entries')
      print(maf.dups[,c("Hugo_Symbol", "Chromosome", "Start_Position", "End_Position", "Tumor_Sample_Barcode", "Variant_Type", "Variant_Classification")])
    }

    maf = maf[!duplicated(maf$id),]
    rownames(maf) = maf$id

    maf = maf[,-ncol(maf)]
    bam_path = as.character(bam_df[bam_df[,1] %in% tsb,2])

    if(length(bam_path) == 0){
      message(paste('Tumor_Sample_Barcode',tsb,'not found in bam_df. Skipping..'))
      maf[,c('t_depth', 't_ref_count', 't_alt_count', 't_vaf')] = NA
      maf.df = rbind(maf.df, maf)
    } else{
      maf.snp = maf[maf$Variant_Type %in% 'SNP',]
      maf.rest = maf[!maf$Variant_Type %in% 'SNP',]

      if(nrow(maf.snp) > 0){

        var = maf.snp[,c('Chromosome', 'Start_Position', 'End_Position', 'Reference_Allele', 'Tumor_Seq_Allele2')]
        write.table(var, 'var_temp.txt', sep = '\t',quote = F, row.names = F, col.names = F)
        cat("Processing ", tsb, " [", nrow(var),  " SNPs] ..\n", sep = '')
        counts = getCounts(var = 'var_temp.txt', bam = bam_path, MapQuality = MapQuality, BaseQuality = BaseQuality, ref_genome = ref_genome)
        maf.snp = merge(maf.snp, counts[,3:6], by = 'row.names', all.x = T)
        maf.snp = maf.snp[,-1]

        if(nrow(maf.rest) > 0 ){
          maf.rest[,c('t_depth', 't_ref_count', 't_alt_count', 't_vaf')] = NA
          maf.df = rbind(maf.df,maf.snp, maf.rest)
        }else{
          maf.df = rbind(maf.df, maf.snp)
        }

      } else{

        cat("No SNPs found in ", tsb, ". Skipping it..\n", sep = '')
        maf[,c('t_depth', 't_ref_count', 't_alt_count', 't_vaf')] = NA
        maf.df = rbind(maf.df, maf)
      }
    }
  }
  message('Done.')
  return(maf.df)
}
