#' Extracts Tumor Sample Barcodes where the given genes are mutated.
#'
#' @description Extracts Tumor Sample Barcodes where the given genes are mutated.
#' @details This function by default looks for field "AAChange" in maf file. One can also manually specify field name containing amino acid changes. Labelling of amino acid changes requires ggrepel package to be installed.
#'
#' @param maf output generated by \code{read.maf}
#' @param genes HGNC symbol for which sample names to be extracted.
#' @param justNames if TRUE, just returns samples names instead of summarized tables.
#' @return list of \code{data.table}s with samples in which given genes are mutated.


genesToBarcodes = function(maf, genes = NULL, justNames = F){

  if(is.null(genes)){
    stop('Atleast one gene name is required.')
  }

  dat = maf$numericMatrix

  if(length(genes[!genes %in% rownames(dat)]) > 0){
    message(paste(genes[!genes %in% rownames(dat)], 'not found in MAF. Excluding it..\n', sep=' '))
    genes = genes[genes %in% rownames(dat)]

    if(length(genes) == 0){
      stop('\nNo genes left !')
    }
  }

  res = list()

  for(i in 1:length(genes)){
    x = dat[genes[i],]
    tsbs = names(x[x!= 0])
    res[i] = list(subset(maf$variant.classification.summary, Tumor_Sample_Barcode %in% tsbs))
    names(res)[i] = genes[i]
  }

  if(justNames){
    res = sapply(res, function(x) return(x$Tumor_Sample_Barcode))
    return(res)
  }else{
    return(res)
  }
}
