#' Draws lollipop plot of amino acid changes on to Protein structure.
#'
#' @description Draws lollipop plot of amino acid changes.
#' @details This function looks for field "AAChange" in maf file. Since there is no hard rule for filed name for amino acid changes, it is advised to manually rename column name to "AAChange" in MAF file if its different.
#' Protein Domain information is obtained from UniProt.
#'
#'
#' @param maf maf maf output generated by \code{read.maf}
#' @param gene HGNC symbol for which protein structure to be drawn.
#' @param label if true lables dots with amino acid conversion using ggrepel package.
#' @return ggplot object of the plot, which can be futher modified.
#' @export


proteinMap = function(maf, gene = NULL, label = F){

  if(is.null(gene)){
    stop('Please provide a gene name.')
  }

  gff = system.file('extdata', 'Uniprot_filtered.txt', package = 'maftools')
  gff = fread(gff, sep = '\t', stringsAsFactors = F, header = T)

  prot = gff[HGNC == gene]

  if(nrow(prot) == 0){
    stop(paste('Structure for protein', gene, 'not found.', sep=' '))
  }

  #Legth of protein
  len = as.numeric(max(prot$End))
  #Remove NA's
  prot = prot[!is.na(prot$Label),]

  #Color code for variant classification
  col = c(brewer.pal(12, name = "Paired"), brewer.pal(11, name = "Spectral")[1:3], "maroon")
  names(col) = names = c("Nonstop_Mutation", "Frame_Shift_Del",
                         "Intron", "Missense_Mutation", "IGR", "Nonsense_Mutation",
                         "RNA", "Splice_Site", "In_Frame_Del", "Frame_Shift_Ins",
                         "Silent", "In_Frame_Ins", "ITD", "3'UTR", "Translation_Start_Site",
                         "two_hit")


  mut = maf$data
  prot.dat = mut[Hugo_Symbol == gene, .(Variant_Type, Variant_Classification, AAChange)]
  if(nrow(prot.dat) == 0){
    stop(paste(gene, 'does not seem to have any mutations!', sep=' '))
  }
  prot.dat = prot.dat[Variant_Classification != 'Splice_Site']
  #Remove 'p.'
  prot.spl = strsplit(x = as.character(prot.dat$AAChange), split = '.', fixed = T)
  prot.conv = sapply(prot.spl, function(x) x[length(x)])

  prot.dat[,conv := prot.conv]
  pos = gsub(pattern = '[[:alpha:]]', replacement = '', x = prot.dat$conv)
  pos = gsub(pattern = '\\*$', replacement = '', x = pos) #Remove * if nonsense mutation ends with *
  pos = as.numeric(sapply(strsplit(x = pos, split = '_', fixed = T), '[[', 1))
  prot.dat[,pos := pos]

  if(nrow( prot.dat[is.na(prot.dat$pos),]) > 0){
    message(paste('Removed', nrow( prot.dat[is.na(prot.dat$pos),]), 'mutations for which AA position was not available', sep = ' '))
    print(prot.dat[is.na(prot.dat$pos),])
    prot.dat = prot.dat[!is.na(prot.dat$pos),]
  }

    # #At present use SNP's
  # prot.snp = prot.dat[Variant_Type == 'SNP']
  # prot.snp$ref = substr(x = prot.snp$conv, start = 1, stop = 1)
  # prot.snp$alt =  substr(x = as.character(prot.snp$conv), start = nchar(prot.snp$conv), stop = nchar(prot.snp$conv))
  # prot.snp$pos = as.numeric(substr(x = as.character(prot.snp$conv), start = 2, stop = nchar(prot.snp$conv)-1))

  prot.snp.sumamry = ddply(prot.dat, .variables = c('Variant_Classification', 'conv', 'pos'), summarise, count =length(AAChange))
  maxCount = max(prot.snp.sumamry$count)

  if(maxCount+1 %% 2 != 0){
    maxCount = maxCount+2
  }

  #Plot points
  p = ggplot()+geom_point(data = prot.snp.sumamry, aes(x = pos, y = count, color = Variant_Classification), size = 2)+scale_color_manual(values = col)+
    theme(legend.position = 'bottom', axis.line.x = element_blank(), legend.title = element_blank())+
    scale_y_continuous(expand = c(0,0), limit = c(0, maxCount))

  #Plot background protein domain
  p = p+geom_rect(data = prot, aes(xmin = 0, xmax = len, ymin = 0.1, ymax = 0.3), fill = 'gray')

  #Plot protein domains. If no domains found, just draw background protein.
  if(nrow(prot) > 0){
    p = p+geom_rect(data = prot, aes(xmin = Start, xmax = End, ymin = 0.1, ymax = 0.3, fill = Label))
  }

  #Plot lines from protein to points (make lollypops)
  p = p+geom_segment(data = prot.snp.sumamry, aes(x = pos, xend = pos, y = 0.3, yend = count ))

  #If user asks to label points, use ggrepel to label.
  if(label){
    require('ggrepel')
    p = p+geom_text_repel(data = prot.snp.sumamry, aes(pos, count, label = as.character(conv)), force = 1)
  }
  print(p)
  return(p)
}
