#' Draws lollipop plot of amino acid changes on to Protein structure.
#'
#' @description Draws lollipop plot of amino acid changes.
#' @details This function by default looks for field "AAChange" in maf file. One can also manually specify field name containing amino acid changes. Labelling of amino acid changes requires ggrepel package to be installed.
#'
#' @param maf output generated by \code{read.maf}
#' @param gene HGNC symbol for which protein structure to be drawn.
#' @param refSeqID RefSeq transcript identifier for \code{gene} if known.
#' @param proteinID RefSeq protein identifier for \code{gene} if known.
#' @param label if \code{TRUE} lables dots with amino acid conversions using \code{ggrepel} package.
#' @param AACol manually specify column name for amino acid changes. Default looks for field 'AAChange'
#' @return ggplot object of the plot, which can be futher modified.
#' @export


lollipopPlot = function(maf, gene = NULL, refSeqID = NULL, proteinID = NULL, label = F, AACol = NULL){

  if(is.null(gene)){
    stop('Please provide a gene name.')
  }

  gff = system.file('extdata', 'protein_domains.txt', package = 'maftools')
  gff = fread(gff, sep = '\t', stringsAsFactors = F, header = T)


  prot = gff[HGNC == gene]

  if(nrow(prot) == 0){
    stop(paste('Structure for protein', gene, 'not found.', sep=' '))
  }

  if(!is.null(refSeqID)){
      prot = gff[refseq.ID == refSeqID]
    } else if(!is.null(proteinID)){
      prot = gff[protein.ID == proteinID]
    } else{
      txs = unique(prot$refseq.ID)
      if(length(txs) > 1){
        message(paste(length(txs), ' transcripts available. Use refSeqID or proteinID to manually specify tx name.', sep = ''))
        print(prot)
        prot = prot[which(prot$aa.length == max(prot$aa.length)),]
        if(length(unique(prot$refseq.ID)) > 1){
          prot = prot[which(prot$refseq.ID == unique(prot[,refseq.ID])[1]),]
          message(paste('Using transcript', unique(prot[,refseq.ID])[1], 'for now.', sep=' '))
        } else{
          message(paste('Using transcript', unique(prot[,refseq.ID])[1], 'for now.', sep=' '))
        }
      }
    }

  #Legth of protein
  len = as.numeric(max(prot$aa.length))
  #Remove NA's
  prot = prot[!is.na(prot$Label),]

  #Color code for variant classification
  col = c(brewer.pal(12, name = "Paired"), brewer.pal(11, name = "Spectral")[1:3], "maroon")
  names(col) = names = c("Nonstop_Mutation", "Frame_Shift_Del",
                         "Intron", "Missense_Mutation", "IGR", "Nonsense_Mutation",
                         "RNA", "Splice_Site", "In_Frame_Del", "Frame_Shift_Ins",
                         "Silent", "In_Frame_Ins", "ITD", "3'UTR", "Translation_Start_Site",
                         "two_hit")


  mut = maf$data

  if(is.null(AACol)){
    if(! 'AAChange' %in% colnames(mut)){
      message('Available fields:')
      print(colnames(mut))
      stop('AAChange field not found in MAF. Use AACol to manually specifiy field name containing protein changes.')
    }
  }else{
    colnames(mut)[which(colnames(mut) == AACol)] = 'AAChange'
  }

  prot.dat = mut[Hugo_Symbol == gene, .(Variant_Type, Variant_Classification, AAChange)]
  if(nrow(prot.dat) == 0){
    stop(paste(gene, 'does not seem to have any mutations!', sep=' '))
  }
  prot.dat = prot.dat[Variant_Classification != 'Splice_Site']
  #Remove 'p.'
  prot.spl = strsplit(x = as.character(prot.dat$AAChange), split = '.', fixed = T)
  prot.conv = sapply(prot.spl, function(x) x[length(x)])

  prot.dat[,conv := prot.conv]
  pos = gsub(pattern = '[[:alpha:]]', replacement = '', x = prot.dat$conv)
  pos = gsub(pattern = '\\*$', replacement = '', x = pos) #Remove * if nonsense mutation ends with *
  pos = as.numeric(sapply(strsplit(x = pos, split = '_', fixed = T), '[[', 1))
  prot.dat[,pos := pos]

  if(nrow( prot.dat[is.na(prot.dat$pos),]) > 0){
    message(paste('Removed', nrow( prot.dat[is.na(prot.dat$pos),]), 'mutations for which AA position was not available', sep = ' '))
    print(prot.dat[is.na(prot.dat$pos),])
    prot.dat = prot.dat[!is.na(prot.dat$pos),]
  }

  prot.snp.sumamry = ddply(prot.dat, .variables = c('Variant_Classification', 'conv', 'pos'), summarise, count =length(AAChange))
  maxCount = max(prot.snp.sumamry$count)


  if(max(prot.snp.sumamry$count) > 10){

    prot.snp.sumamry1 = filter(prot.snp.sumamry, count <= 10)

    prot.snp.sumamry2 = filter(prot.snp.sumamry, count > 10)

    if(nrow(prot.snp.sumamry1) == 0){
      maxCount = 5
    }else{
      maxCount = max(prot.snp.sumamry1$count) + 2
    }

    prot.snp.sumamry$count2 = ifelse(test = prot.snp.sumamry$count > 10, no = prot.snp.sumamry$count, yes = maxCount+2)

    if(nrow(prot.snp.sumamry1) > 0){

      p = ggplot()+geom_point(data = prot.snp.sumamry2, aes(x = pos, y = maxCount+2, color = Variant_Classification), size = 9, alpha = 0.6)+scale_color_manual(values = col)+
        theme(legend.position = 'none', axis.line.x = element_blank(), legend.title = element_blank())+
        scale_y_continuous(breaks = c(0:maxCount, maxCount+3), labels = c(0:maxCount, max(prot.snp.sumamry2$count)+3), limit = c(0, maxCount+3))+xlab('')+ylab('Number of Variants')+scale_x_continuous(limits = c(0, len))+
        geom_text(data = prot.snp.sumamry2, aes(x = pos, y = maxCount+2,label = count))+theme(legend.position = 'none')

      p = p+geom_point(data = prot.snp.sumamry1, aes(x = pos, y = count, color = Variant_Classification), size = 3, alpha = 0.6)+
        theme(legend.position = 'bottom')+guides(colour = guide_legend(override.aes = list(size=3)))
    }else{
      p = ggplot()+geom_point(data = prot.snp.sumamry2, aes(x = pos, y = maxCount+2, color = Variant_Classification), size = 9, alpha = 0.6)+scale_color_manual(values = col)+
        theme(legend.position = 'none', axis.line.x = element_blank(), legend.title = element_blank())+
        scale_y_continuous(breaks = c(0:maxCount, maxCount+3), labels = c(0:maxCount, max(prot.snp.sumamry2$count)+3), limit = c(0, maxCount+3))+xlab('')+ylab('Number of Variants')+scale_x_continuous(limits = c(0, len))+
        geom_text(data = prot.snp.sumamry2, aes(x = pos, y = maxCount+2,label = count))+theme(legend.position = 'none')+
        theme(legend.position = 'bottom')+guides(colour = guide_legend(override.aes = list(size=3)))
    }

  } else{

    maxCount = maxCount+1
    prot.snp.sumamry$count2 = prot.snp.sumamry$count
    #Plot points
    p = ggplot()+geom_point(data = prot.snp.sumamry, aes(x = pos, y = count, color = Variant_Classification), size = 3, alpha = 0.6)+scale_color_manual(values = col)+
      theme(legend.position = 'bottom', axis.line.x = element_blank(), legend.title = element_blank())+
      scale_y_continuous(breaks = c(0:maxCount), labels = c(0:maxCount), limit = c(0, maxCount))+xlab('')+ylab('Number of Variants')

  }

  #Plot background protein domain
  p = p+geom_rect(data = prot, aes(xmin = 0, xmax = len, ymin = 0.1, ymax = 0.3), fill = 'gray')

  #Plot protein domains. If no domains found, just draw background protein.
  if(nrow(prot) > 0){
    p = p+geom_rect(data = prot, aes(xmin = Start, xmax = End, ymin = 0.0, ymax = 0.4, fill = Label))
  }

  #Plot lines from protein to points (make lollypops)
  p = p+geom_segment(data = prot.snp.sumamry, aes(x = pos, xend = pos, y = 0.3, yend = count2-0.03))+ggtitle(label = paste(gene, ' (',unique(prot[,refseq.ID]), ')',sep=''))

  #If user asks to label points, use ggrepel to label.
  if(label){
    require('ggrepel')
    p = p+geom_text_repel(data = prot.snp.sumamry, aes(pos, count2, label = as.character(conv)), force = 2, nudge_y = 0.6, nudge_x = 0.3)
  }
  print(p)
  return(p)
}
