#' Rainfall plot to display kataegis or hyper mutated genomic regions.
#' @description Plots inter variant distance as a function of genomic locus.
#'
#' @param maf an \code{\link{MAF}} object generated by \code{\link{read.maf}}. Required.
#' @param tsb specify sample names (Tumor_Sample_Barcodes) for which plotting has to be done. If NULL, draws plot for most mutated sample.
#' @param color named vector of colors for each coversion class.
#' @param savePlot If TRUE plot is saved to output pdf. Default FALSE.
#' @param width width of plot to be saved.
#' @param height height of plot to be saved.
#' @return returns ggplot object of the plot which can be further modified.
#' @export


rainfallPlot = function(maf, tsb = NULL, color = NULL, savePlot = FALSE, width = 6, height = 3){

  if(is.null(tsb)){
    tsb = as.character(getSampleSummary(maf)[1,Tumor_Sample_Barcode])
  }

  maf.dat = subsetMaf(maf = maf, includeSyn = TRUE, tsb = tsb, fields = 'Hugo_Symbol')
  if(nrow(maf.dat) < 1){
    stop(paste(tsb, 'not found.', sep = ' '))
  }

  maf = maf.dat[,.(Chromosome, Hugo_Symbol, Start_Position, End_Position, Reference_Allele, Tumor_Seq_Allele2, Tumor_Sample_Barcode, Variant_Type)]

  maf.snp = maf[Variant_Type == 'SNP']

  maf.snp$con = paste(maf.snp[,Reference_Allele], maf.snp[,Tumor_Seq_Allele2], sep = '-')

  maf.snp$con.class = suppressWarnings(as.character(factor(maf.snp$con, levels = c("A-G", "T-C", "C-T", "G-A", "A-T", "T-A", "A-C", "T-G", "C-A", "G-T", "C-G", "G-C"),
                                                                   labels = c("A>G", "A>G", "C>T", "C>T", "A>T", "A>T", "A>C", "A>C", "C>A", "C>A", "C>G", "C>G"))))

  maf.snp = transformSegments(maf.snp)
  maf.snp$diff = log10(c(0, diff(maf.snp$Start_Position_updated))+1)

  if(is.null(color)){
    col = RColorBrewer::brewer.pal(n = 6, name = 'Set1')
    names(col) = c('C>T', 'C>G', 'C>A', 'A>T', 'A>G', 'A>C')
  }else{
    col = color
  }

  #hg19 chromosome lengths
  chr.lens = c(249250621, 243199373, 198022430, 191154276, 180915260, 171115067, 159138663,
               146364022, 141213431, 135534747, 135006516, 133851895, 115169878, 107349540,
               102531392, 90354753, 81195210, 78077248, 59128983, 63025520, 48129895, 51304566,
               155270560, 59373566)

  chr.lens.sumsum = cumsum(chr.lens)

  gg.rf = ggplot(data = maf.snp, aes(x= Start_Position_updated, y = diff, col = con.class))+
    geom_point(size = 2, alpha = 0.6)+cowplot::theme_cowplot(font_size = 12)+cowplot::background_grid(major = 'y')+xlab('')+ylab('log10(inter event distance)')+
    theme(axis.line.x = element_blank())+scale_x_continuous(breaks = chr.lens.sumsum, labels = c(1:22, 'X', 'Y'))+
    geom_vline(xintercept = chr.lens.sumsum, linetype = 'dotted', size = 0.3)+theme(legend.position = 'bottom', legend.title = element_blank())+
    scale_color_manual(values = col)+ggtitle(tsb)

  if(savePlot){
    gg.rf = ggplot(data = maf.snp, aes(x= Start_Position_updated, y = diff, col = con.class))+
      geom_point(size = 0.5, alpha = 0.6)+cowplot::theme_cowplot(font_size = 6)+cowplot::background_grid(major = 'y')+xlab('')+ylab('log10(inter event distance)')+
      theme(axis.line.x = element_blank())+scale_x_continuous(breaks = chr.lens.sumsum, labels = c(1:22, 'X', 'Y'))+
      geom_vline(xintercept = chr.lens.sumsum, linetype = 'dotted', size = 0.3)+theme(legend.position = 'bottom', legend.title = element_blank())+
      scale_color_manual(values = col)+ggtitle(tsb)

    cowplot::save_plot(filename = paste(tsb, 'rainfallPlot.pdf', sep = '_'), plot = gg.rf, base_height = height, base_width = width)
  }

  print(gg.rf)
  return(gg.rf)
}
