#' Classifies SNPs into transitions and transversions
#' @description takes output generated by read.maf and classifies Single Nucleotide Variants into Transitions and Transversions.
#'
#' @param maf an MAF object generated by \code{read.maf}
#' @param plot plots a titv fractions. default TRUE.
#' @return list of \code{data.frame}s with Transitions and Transversions summary.
#' @export


titv = function(maf, plot = TRUE)
{
  maf = maf@data
  maf = maf[Variant_Type == 'SNP']

  #Some TCGA studies have Start_Position set to as 'position'. Change if so.
  if(length(grep(pattern = 'Start_position', x = colnames(maf))) > 0){
    colnames(maf)[which(colnames(maf) == 'Start_position')] = 'Start_Position'
  }

  if(length(grep(pattern = 'End_position', x = colnames(maf))) > 0){
    colnames(maf)[which(colnames(maf) == 'Start_position')] = 'End_Position'
  }

  maf = maf[,.(Hugo_Symbol, Start_Position, End_Position, Reference_Allele, Tumor_Seq_Allele2, Tumor_Sample_Barcode)]
  maf$Tumor_Sample_Barcode = as.factor(as.character(maf$Tumor_Sample_Barcode))
  class.summary = c()
  class.summary.df = c()
  TiTv.fractions = data.frame()
  mut.raw.counts = data.frame(row.names = c("A-G", "T-C", "C-T", "G-A", "A-T", "T-A", "A-C", "T-G", "C-A", "G-T", "C-G", "G-C"))

  tsbs = levels(maf[,Tumor_Sample_Barcode])

  for (i in 1:length(tsbs)) {
    mut = data.frame(maf[Tumor_Sample_Barcode == tsbs[i]])

    mut$con = paste(mut[, 4], mut[, 5], sep = "-")
    mut$con = as.factor(mut$con)
    mut$con = factor(mut$con, levels = c("A-G", "T-C", "C-T",
                                         "G-A", "A-T", "T-A", "A-C", "T-G", "C-A", "G-T",
                                         "C-G", "G-C"))
    mut.df = data.frame(absolute_counts = summary(mut$con))
    mut.raw.counts = cbind(mut.raw.counts, mut.df)
    colnames(mut.raw.counts)[ncol(mut.raw.counts)] = tsbs[i]
    mut.df$class = rep(rownames(mut.df)[seq(1, 12, by = 2)],each = 2)
    mut.df$TiTv = c(rep(x = "Ti", 4), rep("Tv", 8))
    mut.df$percent_contribution = mut.df$absolute_counts/sum(mut.df$absolute_counts) * 100
    mut.df$conversion = row.names(mut.df)
    mut.df = mut.df[, c(5, 1:4)]
    mut.summarized = ddply(.data = mut.df, .variables = ~class, summarise, sum(percent_contribution))
    colnames(mut.summarized) = c("class", "percent_contribution")
    mut.summarized2 = ddply(.data = mut.df, .variables = ~TiTv,
                            summarise, sum(percent_contribution))
    colnames(mut.summarized2) = c("TiTv", "percent_contribution")
    TiTv.fractions = rbind(TiTv.fractions, mut.summarized2$percent_contribution)
    rownames(TiTv.fractions)[i] = tsbs[i]
    mut.summary = list(mut.df, mut.summarized, mut.summarized2)
    class.summary.df = rbind(class.summary.df, c(mut.summary[[2]][,2], tsbs[i]))
    class.summary = rbind(class.summary, cbind(mut.summary[[2]],
                                               pid = tsbs[i]))
  }
  #add titv classification
  class.summary$TiTv = suppressWarnings( as.character(factor(x = class.summary$class, levels = mut.df$class, labels = mut.df$TiTv)) )
  colnames(TiTv.fractions) = c("Ti", "Tv")
  if (plot) {
    p = ggplot(data = class.summary, aes(x = class, y = percent_contribution, color = TiTv)) +
      geom_boxplot() + theme_minimal() + ylim(0, 100) +
      xlab("Conversion Class") + theme(axis.line = element_line(colour = "black")) +
      ylab("Fraction of Mutations")+scale_color_manual(values = c('Ti' = 'maroon', 'Tv' = 'royalblue'))
    print(p)
  }
  class.summary.df = as.data.frame(class.summary.df[, c(7, 1:6)])
  colnames(class.summary.df) = c("Tumor_Sample_Barcode", mut.summary[[2]][,1])
  class.summary.df[,2:7] = apply(class.summary.df[,2:7], 2, as.numeric)
  return(list(fraction.contribution = class.summary.df, raw.counts = mut.raw.counts, TiTv.fractions = TiTv.fractions))
}
